<?xml version="1.0"?>
<robot name="bot1">

  <link name="world" />

  <material name="black">
    <color rgba="0 0 0 1" />
  </material>

  <material name="blue">
    <color rgba="0 0 0.8 1" />
  </material>

  <material name="white">
    <color rgba="1 1 1 1" />
  </material>

  <link name="base_link">
    <inertial>
      <mass value="5.0" />
      <origin xyz="0 0 0.05" rpy="0 0 0" />
      <inertia ixx="0.1" ixy="0.0" ixz="0.0" iyy="0.1" iyz="0.0" izz="0.1" />
    </inertial>

    <visual>
      <geometry>
        <cylinder length="0.1" radius="0.5" />
      </geometry>
      <origin xyz="0 0 0.05" rpy="0 0 0" />
    </visual>

    <collision>
      <geometry>
        <cylinder length="0.1" radius="0.5" />
      </geometry>
      <origin xyz="0 0 0.05" rpy="0 0 0" />
    </collision>


  </link>

  <link name="arm1">
    <inertial>
      <mass value="2.0" />
      <origin xyz="0 0 0.4" rpy="0 0 0" />
      <inertia ixx="0.02" ixy="0.0" ixz="0.0" iyy="0.02" iyz="0.0" izz="0.005" />
    </inertial>

    <visual>
      <geometry>
        <cylinder length="0.8" radius="0.2" />
      </geometry>
      <origin xyz="0 0 0.4" rpy="0 0 0" />
      <material name="white" />
    </visual>

    <collision>
      <geometry>
        <cylinder length="0.8" radius="0.2" />
      </geometry>
      <origin xyz="0 0 0.4" rpy="0 0 0" />
    </collision>

  </link>

  <joint name="base_to_world" type="fixed">
    <parent link="world" />
    <child link="base_link" />
    <origin xyz="0 0 0" />
  </joint>

  <joint name="bodytoarm1" type="revolute">
    <parent link="base_link" />
    <child link="arm1" />
    <origin xyz="0 0 0.1" rpy="0 0 0" />
    <axis xyz="0 0 1" />
    <limit lower="-3.14" upper="3.14" effort="20.0" velocity="1.0" />
  </joint>


  <link name="arm2">
    <inertial>
      <mass value="1.5" />
      <origin xyz="0 0 0.4" rpy="0 0 0" />
      <inertia ixx="0.015" ixy="0.0" ixz="0.0" iyy="0.015" iyz="0.0" izz="0.004" />
    </inertial>

    <visual>
      <geometry>
        <cylinder length="0.8" radius="0.2" />
      </geometry>
      <origin xyz="0 0 0.4" rpy="0 0 0" />
      <material name="blue" />
    </visual>

    <collision>
      <geometry>
        <cylinder length="0.8" radius="0.2" />
      </geometry>
      <origin xyz="0 0 0.4" rpy="0 0 0" />
    </collision>

  </link>


  <joint name="arm1toarm2" type="revolute">
    <parent link="arm1" />
    <child link="arm2" />
    <origin xyz="0 0 0.8" rpy="0 0 0" />
    <axis xyz="0 1 0" />
    <limit lower="-1.57" upper="1.57" effort="15.0" velocity="1.0" />
  </joint>


  <link name="arm3">

    <inertial>
      <mass value="1.0" />
      <origin xyz="0 0 0.4" rpy="0 0 0" />
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.002" />
    </inertial>

    <visual>
      <geometry>
        <cylinder length="0.8" radius="0.2" />
      </geometry>
      <origin xyz="0 0 0.4" rpy="0 0 0" />
      <material name="black" />
    </visual>

    <collision>
      <geometry>
        <cylinder length="0.8" radius="0.2" />
      </geometry>
      <origin xyz="0 0 0.4" rpy="0 0 0" />
    </collision>

  </link>

  <joint name="arm2toarm3" type="revolute">
    <parent link="arm2" />
    <child link="arm3" />
    <origin xyz="0 0 0.8" rpy="0 0 0" />
    <axis xyz="1 0 0" />
    <limit lower="-1.57" upper="1.57" effort="10.0" velocity="1.0" />
  </joint>

  <ros2_control name="GazeboSystem" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>

    <joint name="bodytoarm1">
      <command_interface name="position">
        <param name="min">-3.142</param>
        <param name="max">3.142</param>
      </command_interface>
      <state_interface name="position" />
      <state_interface name="velocity" />
      <state_interface name="effort" />
    </joint>

    <joint name="ar1toarm2">
      <command_interface name="position">
        <param name="min">-1.57</param>
        <param name="max">1.57</param>
      </command_interface>
      <state_interface name="position" />
      <state_interface name="velocity" />
      <state_interface name="effort" />
    </joint>

    <joint name="rev_3">
      <command_interface name="position">
        <param name="min">-1.57</param>
        <param name="max">1.57</param>
      </command_interface>
      <state_interface name="position" />
      <state_interface name="velocity" />
      <state_interface name="effort" />
    </joint>
  </ros2_control>

  <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
      <parameters>/home/abhinand/Documents/3arm_robotic_manipulator/DC/config/robot_config.yaml</parameters>
    </plugin>
  </gazebo>
</robot>